File: ScaleformUI_Csharp/Scaleforms/PauseMenu/PauseMenuScaleform.cs
using CitizenFX.Core;
using ScaleformUI.Elements;
using ScaleformUI.Menu;
using ScaleformUI.Menus;
using System.Drawing;
using static CitizenFX.Core.Native.API;

namespace ScaleformUI.Scaleforms
{
    public class PauseMenuScaleform
    {
        public ScaleformWideScreen _header;
        public ScaleformWideScreen _pause;
        public ScaleformWideScreen _lobby;
        public ScaleformWideScreen _pauseBG;
        internal bool BGEnabled;
        private bool _visible;
        internal bool Loaded => _header is not null && _header.IsLoaded && _pause is not null && _pause.IsLoaded && _lobby is not null && _lobby.IsLoaded && _pauseBG is not null && _pauseBG.IsLoaded;
        public bool Visible { get => _visible; set => _visible = value; }

        public PauseMenuScaleform()
        {
        }

        public void Load()
        {
            _header = new ScaleformWideScreen("pausemenuheader");
            _pause = new ScaleformWideScreen("ScaleformUIPause");
            //_lobby = new ScaleformWideScreen("lobbymenu");
            //_pauseBG = new ScaleformWideScreen("store_background");
        }

        internal void FadeInMenus()
        {
            _header.CallFunction("DRAW_MENU");
            _pause.CallFunction("DRAW_MENU");
            //_lobby.CallFunction("DRAW_MENU");
        }

        public void SetDataSlotEmpty(PM_COLUMNS column)
        {
            _pause.CallFunction("SET_DATA_SLOT_EMPTY", (int)column);
        }

        public void SetHeaderTitle(string title, string subtitle = "", bool shiftUpHeader = false)
        {
            _header.CallFunction("SET_HEADER_TITLE", title, subtitle, shiftUpHeader);
        }

        public void SetHeaderDetails(string topDetail, string midDetail, string botDetail)
        {
            _header.CallFunction("SET_HEADER_DETAILS", topDetail, midDetail, botDetail, false);
        }

        public void ShiftCoronaDescription(bool shiftDesc, bool hideTabs)
        {
            _header.CallFunction("SHIFT_CORONA_DESC", shiftDesc, hideTabs);
        }

        public void ShowHeadingDetails(bool show)
        {
            _header.CallFunction("SHOW_HEADING_DETAILS", show);
        }

        public void SetHeaderCharImg(string txd, string charTexturePath, bool show)
        {
            _header.CallFunction("SET_HEADER_CHAR_IMG", txd, charTexturePath, show);
        }

        public void SetHeaderSecondaryImg(string txd, string charTexturePath, bool show)
        {
            _header.CallFunction("SET_HEADER_CREW_IMG", txd, charTexturePath, show);
        }

        public void HeaderGoRight()
        {
            _header.CallFunction("GO_RIGHT");
        }
        public void HeaderGoLeft()
        {
            _header.CallFunction("GO_LEFT");
        }

        public void AddPauseMenuTab(string title, int tabType, int tabContentType, SColor color)
        {
            _header.CallFunction("ADD_HEADER_TAB", title, tabType, color);
        }
        public void AddLobbyMenuTab(string title, int tabType, SColor color)
        {
            _header.CallFunction("ADD_HEADER_TAB", title, tabType, color);
        }

        public void SelectTab(int tab)
        {
            _header.CallFunction("SET_TAB_INDEX", tab);
        }

        public void SetFocus(int focus, bool bDontFallOff = false, bool bSkipInputSpamCheck = false)
        {
            _pause.CallFunction("MENU_SHIFT_DEPTH", focus, bDontFallOff, bSkipInputSpamCheck);
        }

        public void AddLeftItem(int type, string title, SColor itemColor, SColor highlightColor, bool enabled = true)
        {
            if (itemColor != SColor.HUD_None && highlightColor != SColor.HUD_None)
                _pause.CallFunction("ADD_LEFT_ITEM", type, title, enabled, itemColor, highlightColor);
            else if (itemColor != SColor.HUD_None)
                _pause.CallFunction("ADD_LEFT_ITEM", type, title, enabled, itemColor);
            else
                _pause.CallFunction("ADD_LEFT_ITEM", type, title, enabled);
        }

        public void AddRightTitle(int leftItem, string title)
        {
            _pause.CallFunction("ADD_RIGHT_TITLE", leftItem, title);
        }

        public void AddRightListLabel(int leftItem, string label, string fontName, int fontId)
        {
            AddTextEntry($"PauseMenu_{leftItem}", label);
            BeginScaleformMovieMethod(_pause.Handle, "ADD_RIGHT_LIST_ITEM");
            ScaleformMovieMethodAddParamInt(leftItem);
            ScaleformMovieMethodAddParamInt(0);
            BeginTextCommandScaleformString($"PauseMenu_{leftItem}");
            EndTextCommandScaleformString_2();
            ScaleformMovieMethodAddParamPlayerNameString(fontName);
            ScaleformMovieMethodAddParamInt(fontId);
            EndScaleformMovieMethod();
        }

        public void AddRightStatItemLabel(int leftItem, string label, string rightLabel, ItemFont labelFont, ItemFont rLabelFont)
        {
            _pause.CallFunction("ADD_RIGHT_LIST_ITEM", leftItem, 1, 0, label, rightLabel, -1, labelFont.FontName, labelFont.FontID, rLabelFont.FontName, rLabelFont.FontID);
        }

        public void AddRightStatItemColorBar(int leftItem, string label, int value, SColor barColor, ItemFont labelFont)
        {
            _pause.CallFunction("ADD_RIGHT_LIST_ITEM", leftItem, 1, 1, label, value, barColor, labelFont.FontName, labelFont.FontID);
        }

        public void AddRightSettingsBaseItem(int leftItem, string label, string rightLabel, bool enabled)
        {
            _pause.CallFunction("ADD_RIGHT_LIST_ITEM", leftItem, 2, 0, label, enabled, rightLabel);
        }

        public void AddRightSettingsListItem(int leftItem, string label, List<dynamic> items, int startIndex, bool enabled)
        {
            string stringList = string.Join(",", items);
            _pause.CallFunction("ADD_RIGHT_LIST_ITEM", leftItem, 2, 1, label, enabled, stringList, startIndex);
        }

        public void AddRightSettingsProgressItem(int leftItem, string label, int max, SColor color, int index, bool enabled)
        {
            _pause.CallFunction("ADD_RIGHT_LIST_ITEM", leftItem, 2, 2, label, enabled, max, color, index);
        }
        public void AddRightSettingsProgressItemAlt(int leftItem, string label, int max, SColor color, int index, bool enabled)
        {
            _pause.CallFunction("ADD_RIGHT_LIST_ITEM", leftItem, 2, 3, label, enabled, max, color, index);
        }

        public void AddRightSettingsSliderItem(int leftItem, string label, int max, SColor color, int index, bool enabled)
        {
            _pause.CallFunction("ADD_RIGHT_LIST_ITEM", leftItem, 2, 5, label, enabled, max, color, index);
        }
        public void AddRightSettingsCheckboxItem(int leftItem, string label, UIMenuCheckboxStyle style, bool check, bool enabled)
        {
            _pause.CallFunction("ADD_RIGHT_LIST_ITEM", leftItem, 2, 4, label, enabled, (int)style, check);
        }

        public void AddKeymapTitle(int leftItem, string title, string rightLabel_1, string rightLabel_2)
        {
            _pause.CallFunction("ADD_RIGHT_TITLE", leftItem, title, rightLabel_1, rightLabel_2);
        }
        public void AddKeymapItem(int leftItem, string label, string control1, string control2)
        {
            BeginScaleformMovieMethod(_pause.Handle, "ADD_RIGHT_LIST_ITEM");
            ScaleformMovieMethodAddParamInt(leftItem);
            ScaleformMovieMethodAddParamInt(3);
            PushScaleformMovieFunctionParameterString(label);
            BeginTextCommandScaleformString("string");
            AddTextComponentScaleform(control1);
            EndTextCommandScaleformString_2();
            BeginTextCommandScaleformString("string");
            AddTextComponentScaleform(control2);
            EndTextCommandScaleformString_2();
            EndScaleformMovieMethod();
        }
        public void UpdateKeymap(int leftItem, int rightItem, string control1, string control2)
        {
            BeginScaleformMovieMethod(_pause.Handle, "UPDATE_KEYMAP_ITEM");
            ScaleformMovieMethodAddParamInt(leftItem);
            ScaleformMovieMethodAddParamInt(rightItem);
            BeginTextCommandScaleformString("string");
            AddTextComponentScaleform(control1);
            EndTextCommandScaleformString_2();
            BeginTextCommandScaleformString("string");
            AddTextComponentScaleform(control2);
            EndTextCommandScaleformString_2();
            EndScaleformMovieMethod();
        }

        public void SetRightSettingsItemBool(int leftItem, int rightItem, bool value)
        {
            _pause.CallFunction("SET_RIGHT_SETTINGS_ITEM_VALUE", leftItem, rightItem, value);
        }
        public void SetRightSettingsItemIndex(int leftItem, int rightItem, int value)
        {
            _pause.CallFunction("SET_RIGHT_SETTINGS_ITEM_VALUE", leftItem, rightItem, value);
        }
        public void SetRightSettingsItemValue(int leftItem, int rightItem, int value)
        {
            _pause.CallFunction("SET_RIGHT_SETTINGS_ITEM_VALUE", leftItem, rightItem, value);
        }

        public void UpdateItemRightLabel(int leftItem, int rightItem, string label)
        {
            _pause.CallFunction("UPDATE_RIGHT_ITEM_RIGHT_LABEL", leftItem, rightItem, label);
        }

        public void UpdateStatsItem(int leftItem, int rightItem, string label, string rightLabel)
        {
            _pause.CallFunction("UPDATE_RIGHT_STATS_ITEM", leftItem, rightItem, label, rightLabel);
        }
        public void UpdateStatsItem(int leftItem, int rightItem, string label, int value, SColor color)
        {
            _pause.CallFunction("UPDATE_RIGHT_STATS_ITEM", leftItem, rightItem, label, value, color);
        }

        public void UpdateItemColoredBar(int leftItem, int rightItem, SColor color)
        {
            if (color == SColor.HUD_None)
                _pause.CallFunction("UPDATE_COLORED_BAR_COLOR", leftItem, rightItem, (int)HudColor.HUD_COLOUR_WHITE);
            else
                _pause.CallFunction("UPDATE_COLORED_BAR_COLOR", leftItem, rightItem, color);
        }
        public async Task<string> SendInputEvent(int direction)
        {
            BeginScaleformMovieMethod(_pause.Handle, "SET_INPUT_EVENT");
            ScaleformMovieMethodAddParamInt(direction);
            int ret = EndScaleformMovieMethodReturnValue();
            while (!IsScaleformMovieMethodReturnValueReady(ret)) await BaseScript.Delay(0);
            string res = GetScaleformMovieFunctionReturnString(ret);
            return res;
        }

        public void SendScrollEvent(int direction)
        {
            _pause.CallFunction("SET_SCROLL_EVENT", direction);
        }
        public async Task<string> SendClickEvent()
        {
            BeginScaleformMovieMethod(_pause.Handle, "MOUSE_CLICK_EVENT");
            int ret = EndScaleformMovieMethodReturnValue();
            while (!IsScaleformMovieMethodReturnValueReady(ret)) await BaseScript.Delay(0);
            string res = GetScaleformMovieFunctionReturnString(ret);
            return res;
        }

        public void Dispose()
        {
            _visible = false;
            _pause?.CallFunction("CLEAR_ALL");
            //_lobby?.CallFunction("CLEAR_ALL");
            _header?.CallFunction("CLEAR_ALL");
            _pause?.Dispose();
            //_lobby?.Dispose();
            _header?.Dispose();
            firstTick = true;
        }

        bool firstTick = true;
        public void Draw(bool isLobby = false)
        {
            if (_visible)
            {
                if (IsFrontendReadyForControl())
                {
                    SetScriptGfxDrawBehindPausemenu(true);
                    if (firstTick)
                    {
                        FadeInMenus();
                        firstTick = false;
                    }

                    Vector2 headerPos = new Vector2(0.501f, 0.162f);
                    SizeF headerSize = new SizeF(0.6782f, 0.145f);

                    Vector2 pausePos = new Vector2(0.6617187f, 0.7226667f);
                    SizeF pauseSize = new SizeF(1, 1);

                    ScreenTools.AdjustNormalized16_9ValuesForCurrentAspectRatio(3, ref headerPos, ref headerSize);
                    ScreenTools.AdjustNormalized16_9ValuesForCurrentAspectRatio(3, ref pausePos, ref pauseSize);

                    if (BGEnabled)
                        _pauseBG.Render2D();
                    DrawScaleformMovie(_header.Handle, headerPos.X, headerPos.Y, headerSize.Width, headerSize.Height, 255, 255, 255, 255, 0);
                    DrawScaleformMovie(_pause.Handle, pausePos.X, pausePos.Y, pauseSize.Width, pauseSize.Height, 255, 255, 255, 255, 0);

                    //if (!isLobby)
                    //else
                    //    DrawScaleformMovie(_lobby.Handle, pausePos.X, pausePos.Y, pauseSize.Width, pauseSize.Height, 255, 255, 255, 255, 0);
                }
            }
        }
    }
}

File: ScaleformUI_Lua/src/scaleforms/PauseMenu/PauseMenuScaleform.lua
PauseMenu = setmetatable({}, PauseMenu)
PauseMenu.__index = PauseMenu
PauseMenu.__call = function()
    return "PauseMenu"
end

---@class PauseMenu: Scaleform
---@field public _header Scaleform
---@field public _pause Scaleform
---@field public _pauseBG Scaleform
---@field public Loaded boolean
---@field public _visible boolean
---@field public BGEnabled boolean
---@field public Visible fun(self:PauseMenu, visible:boolean):boolean
---@field public SetHeaderTitle fun(self:PauseMenu, title:string, subtitle:string | nil, shiftUpHeader:boolean | nil):nil
---@field public SetHeaderDetails fun(self:PauseMenu, topDetail:string, midDetail:string, botDetail:string):nil
---@field public ShiftCoronaDescription fun(self:PauseMenu, shiftDesc:boolean, hideTabs:boolean):nil
---@field public ShowHeaderDetails fun(self:PauseMenu, show:boolean):nil
---@field public SetHeaderCharImg fun(self:PauseMenu, txd:string, charTexturePath:string, show:boolean):nil
---@field public SetHeaderSecondaryImg fun(self:PauseMenu, txd:string, charTexturePath:string, show:boolean):nil
---@field public HeaderGoRight fun(self:PauseMenu):nil
---@field public HeaderGoLeft fun(self:PauseMenu):nil
---@field public AddPauseMenuTab fun(self:PauseMenu, title:string, _type:number, _tabContentType:number, color:SColor | nil):nil
---@field public AddLobbyMenuTab fun(self:PauseMenu, title:string, _type:number, _tabContentType:number, color:SColor | nil):nil
---@field public SelectTab fun(self:PauseMenu, tab:number):nil
---@field public SetFocus fun(self:PauseMenu, focus:number):nil
---@field public AddLeftItem fun(self:PauseMenu, _type:number, title:string, _tabContentType:number, color:SColor, enabled:boolean):nil
---@field public AddRightTitle fun(self:PauseMenu, leftItemIndex:number, title:string):nil
---@field public AddRightListLabel fun(self:PauseMenu, leftItemIndex:number, title:string, fontName:string, fontId:number):nil
---@field public AddRightStatItemLabel fun(self:PauseMenu, leftItemIndex:number, label:string, rightLabel:string, labelFont:Font, rLabelFont:Font):nil
---@field public AddRightStatItemColorBar fun(self:PauseMenu, leftItemIndex:number, label:string, value:number, color:SColor, labelFont:Font):nil
---@field public AddRightSettingsBaseItem fun(self:PauseMenu, leftItemIndex:number, label:string, rightLabel:string, enabled:boolean):nil
---@field public AddRightSettingsListItem fun(self:PauseMenu, leftItemIndex:number, label:string, items:table, startIndex:number, enabled:boolean):nil
---@field public AddRightSettingsProgressItem fun(self:PauseMenu, leftItemIndex:number, label:string, max:number, colour:SColor, index:number, enabled:boolean):nil
---@field public AddRightSettingsProgressItemAlt fun(self:PauseMenu, leftItemIndex:number, label:string, max:number, colour:SColor, index:number, enabled:boolean):nil
---@field public AddRightSettingsSliderItem fun(self:PauseMenu, leftItemIndex:number, label:string, max:number, colour:SColor, index:number, enabled:boolean):nil
---@field public AddRightSettingsCheckboxItem fun(self:PauseMenu, leftItemIndex:number, label:string, style:number, check:boolean, enabled:boolean):nil
---@field public AddKeymapTitle fun(self:PauseMenu, leftItemIndex:number, title:string, rightLabel:string, rightLabel2:string):nil
---@field public AddKeymapItem fun(self:PauseMenu, leftItemIndex:number, label:string, control:string, control2:string):nil
---@field public UpdateKeymap fun(self:PauseMenu, leftItemIndex:number, rightIndex:number, control:string, control2:string):nil
---@field public SetRightSettingsItemBool fun(self:PauseMenu, leftItemIndex:number, rightItemIndex:number, check:boolean):nil
---@field public SetRightSettingsItemIndex fun(self:PauseMenu, leftItemIndex:number, rightItemIndex:number, value:number):nil
---@field public SetRightSettingsItemValue fun(self:PauseMenu, leftItemIndex:number, rightItemIndex:number, value:number):nil
---@field public UpdateItemRightLabel fun(self:PauseMenu, leftItemIndex:number, rightItemIndex:number, rightLabel:string):nil
---@field public UpdateStatsItemBasic fun(self:PauseMenu, leftItemIndex:number, rightItemIndex:number, label:string, rightLabel:string):nil
---@field public UpdateStatsItemBar fun(self:PauseMenu, leftItemIndex:number, rightItemIndex:number, label:string, value:number, color:SColor):nil
---@field public UpdateItemColoredBar fun(self:PauseMenu, leftItemIndex:number, rightItemIndex:number, colour:SColor):nil
---@field public SendInputEvent fun(self:PauseMenu, inputEvent:string):nil
---@field public SendScrollEvent fun(self:PauseMenu, scrollEvent:number):nil
---@field public SendClickEvent fun(self:PauseMenu):nil
---@field public Dispose fun(self:PauseMenu):nil
---@field public Draw fun(self:PauseMenu, isLobby:boolean | nil):nil
---@field public Load fun(self:PauseMenu):nil
---@field public IsLoaded fun(self:PauseMenu):boolean

---Creates a new instance of the pause menu
---@return PauseMenu
function PauseMenu.New()
    local _data = {
        _header = nil,
        _pause = nil,
        _pauseBG = nil,
        BGEnabled = false,
        Loaded = false,
        _visible = false,
        firstTick = true,
    }
    return setmetatable(_data, PauseMenu)
end

---Toggle the visibility of the pause menu
---@param visible boolean
---@return boolean
function PauseMenu:Visible(visible)
    if ToBool(visible) then
        self._visible = visible
    end
    return self._visible
end

---Load the pause menu scaleforms
function PauseMenu:Load()
    if (self._header ~= nil and self._pause ~= nil and self._pauseBG ~= nil) then return end
    self._header = Scaleform.RequestWidescreen("pausemenuheader")
    self._pause = Scaleform.RequestWidescreen("ScaleformUIPause")
    self._pauseBG = Scaleform.RequestWidescreen("store_background")
end

function PauseMenu:FadeInMenus()
    self._header:CallFunction("DRAW_MENU")
    self._pause:CallFunction("DRAW_MENU")
end

function PauseMenu:IsLoaded()
    return self._header ~= nil and self._header:IsLoaded() and self._pause ~= nil and self._pause:IsLoaded() and self._pauseBG ~= nil and self._pauseBG:IsLoaded()
end

---Set the header title and subtitle text of the pause menu header
---@param title string
---@param subtitle string | nil
---@param shiftUpHeader boolean | nil
function PauseMenu:SetHeaderTitle(title, subtitle, shiftUpHeader)
    if (subtitle == nil) then subtitle = "" end
    if (shiftUpHeader == nil) then shiftUpHeader = false end
    self._header:CallFunction("SET_HEADER_TITLE", title, subtitle, shiftUpHeader)
end

---Set the header details of the pause menu header
---@param topDetail string
---@param midDetail string
---@param botDetail string
function PauseMenu:SetHeaderDetails(topDetail, midDetail, botDetail)
    self._header:CallFunction("SET_HEADER_DETAILS", topDetail, midDetail, botDetail, false)
end

---Shift the corona description of the pause menu header
---@param shiftDesc boolean
---@param hideTabs boolean
function PauseMenu:ShiftCoronaDescription(shiftDesc, hideTabs)
    self._header:CallFunction("SHIFT_CORONA_DESC", shiftDesc, hideTabs)
end

---Toggle the header details of the pause menu header
---@param show boolean
function PauseMenu:ShowHeadingDetails(show)
    self._header:CallFunction("SHOW_HEADING_DETAILS", show)
end

---Set the header character headshot of the pause menu header
---@param txd string
---@param charTexturePath string
---@param show boolean
function PauseMenu:SetHeaderCharImg(txd, charTexturePath, show)
    self._header:CallFunction("SET_HEADER_CHAR_IMG", txd, charTexturePath, show)
end

---Set the header secondary image of the pause menu header
---@param txd string
---@param charTexturePath string
---@param show boolean
function PauseMenu:SetHeaderSecondaryImg(txd, charTexturePath, show)
    self._header:CallFunction("SET_HEADER_CREW_IMG", txd, charTexturePath, show)
end

---Move selection of the header to the right
function PauseMenu:HeaderGoRight()
    self._header:CallFunction("GO_RIGHT")
end

---Move selection of the header to the left
function PauseMenu:HeaderGoLeft()
    self._header:CallFunction("GO_LEFT")
end

---Add a tab to the pause menu
---@param title string
---@param _type number
---@param color SColor Sets the color of the tab (default: SColor.HUD_Freemode)
function PauseMenu:AddPauseMenuTab(title, _type, color)
    if color == nil then color = SColor.HUD_Freemode end
    self._header:CallFunction("ADD_HEADER_TAB", title, _type, color)
end

---Select a tab in the pause menu
function PauseMenu:SelectTab(tab)
    self._header:CallFunction("SET_TAB_INDEX", tab)
end

---Set the focued item of the pause menu
---@param focus number
function PauseMenu:SetFocus(focus, bDontFallOff, bSkipInputSpamCheck)
    if bDontFallOff == nil then bDontFallOff = false end
    if  bSkipInputSpamCheck == nil then  bSkipInputSpamCheck = false end
    self._pause:CallFunction("MENU_SHIFT_DEPTH", focus, bDontFallOff, bSkipInputSpamCheck)
end

---Dispose the pause menu
function PauseMenu:Dispose()
    self._visible = false
    self._pause:CallFunction("CLEAR_ALL")
    self._header:CallFunction("CLEAR_ALL")
    -- self._pause:Dispose()
    -- self._header:Dispose()
    -- self._pauseBG:Dispose()
    self.firstTick = true;
end

---Draw the pause menu
---@param isLobby boolean|nil
function PauseMenu:Draw(isLobby)
    if isLobby == nil then isLobby = false end
    if self._visible then
        if IsFrontendReadyForControl() then
            SetScriptGfxDrawBehindPausemenu(true)
            if IsUsingKeyboard(2) then
                SetMouseCursorActiveThisFrame()
            end
            if self.firstTick then
                self:FadeInMenus()
                self.firstTick = false;
            end
            if self.BGEnabled then
                self._pauseBG:Render2D()
            end

            local headx, heady, headw, headh = AdjustNormalized16_9ValuesForCurrentAspectRatio(0.501, 0.162, 0.6782, 0.145)
            local pausex, pausey, pausew, pauseh = AdjustNormalized16_9ValuesForCurrentAspectRatio(0.6617187, 0.7226667, 1.0, 1.0)

            self._header:Render2DNormal(headx, heady, headw, headh)
            self._pause:Render2DNormal(pausex, pausey, pausew, pauseh)
        end
    end
end
